File: day6/part2.py
from typing import List
import math


def read_input(path: str) -> List[str]:
    with open(path, "r", encoding="utf-8") as f:
        return [line.rstrip("\n") for line in f]


def split_blocks(lines: List[str]) -> List[List[str]]:
    width = max(len(line) for line in lines)
    padded = [line.ljust(width) for line in lines]

    used = [any(row[c] != " " for row in padded) for c in range(width)]

    blocks = []
    start = None

    for i, active in enumerate(used + [False]):
        if active and start is None:
            start = i
        elif not active and start is not None:
            blocks.append([row[start:i] for row in padded])
            start = None

    return blocks


def solve_block(block: List[str]) -> int:
    height = len(block)
    width = len(block[0])

    operator = None
    numbers: List[int] = []

    for col in range(width - 1, -1, -1):
        digits = []
        for row in range(height):
            ch = block[row][col]
            if ch.isdigit():
                digits.append(ch)
            elif ch in "+*":
                operator = ch

        if digits:
            numbers.append(int("".join(digits)))

    if operator == "+":
        return sum(numbers)
    if operator == "*":
        return math.prod(numbers)

    raise ValueError("Operator not found")


def main() -> None:
    lines = read_input("input.txt")
    blocks = split_blocks(lines)
    total = sum(solve_block(b) for b in blocks)
    print(total)


if __name__ == "__main__":
    main()

File: day6/part2.ts
import * as fs from "fs";

function readInput(path: string): string[] {
  return fs.readFileSync(path, "utf8").split("\n");
}

function splitBlocks(lines: string[]): string[][] {
  const width = Math.max(...lines.map((l) => l.length));
  const padded = lines.map((l) => l.padEnd(width, " "));

  const used = Array(width).fill(false);
  for (let c = 0; c < width; c++) {
    for (const row of padded) {
      if (row[c] !== " ") {
        used[c] = true;
        break;
      }
    }
  }

  const blocks: string[][] = [];
  let start: number | null = null;

  for (let i = 0; i <= width; i++) {
    const active = i < width && used[i];
    if (active && start === null) start = i;
    else if (!active && start !== null) {
      blocks.push(padded.map((r) => r.slice(start!, i)));
      start = null;
    }
  }

  return blocks;
}

function solveBlock(block: string[]): bigint {
  const h = block.length;
  const w = block[0].length;

  const numbers: bigint[] = [];
  let op: string | null = null;

  for (let c = w - 1; c >= 0; c--) {
    let digits = "";
    for (let r = 0; r < h; r++) {
      const ch = block[r][c];
      if (/\d/.test(ch)) digits += ch;
      else if (ch === "+" || ch === "*") op = ch;
    }
    if (digits) numbers.push(BigInt(digits));
  }

  return op === "+"
    ? numbers.reduce((a, b) => a + b, 0n)
    : numbers.reduce((a, b) => a * b, 1n);
}

const lines = readInput("input.txt");
const blocks = splitBlocks(lines);
const total = blocks.reduce((s, b) => s + solveBlock(b), 0n);
console.log(total.toString());
